# 134B Week9

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

Assume that the current exchange rate is
$0.90/=C.

![table](../img/134B_Week9_1.png "Title")

## Exercise 8.1
Suppose that 1- and 2-year oil forward prices are $22/barrel and $23/barrel. The 1-
and 2-year interest rates are 6% and 6.5%. Show that the new 2-year swap price is
$22.483.

### Solution:
We first compute the present value of the payment:

In [2]:
r1 = 0.06
r2 = 0.065
PV = 22*(1+r1)**-1 + 23*(1+r2)**-2
print(f'The present value of the cash flows is ${PV:.2f}')

The present value of the cash flows is $41.03


Then we solve for $x$:
$$ \frac{x}{1+r_1} + \frac{x}{(1+r_2)^2} = PV$$

In [3]:
x = PV*((1+r1)**-1 + (1+r2)**-2)**-1
print(f'The two year swap price is ${x:.2f}')

The two year swap price is $22.48


## Exercise 8.2
Suppose that oil forward prices for 1 year, 2 years, and 3 years are $20, $21, and $22.
The 1-year effective annual interest rate is 6.0%, the 2-year interest rate is 6.5%, and
the 3-year interest rate is 7.0%.
- a. What is the 3-year swap price?
- b. What is the price of a 2-year swap beginning in one year? (That is, the first
swap settlement will be in 2 years and the second in 3 years.)

### Solution
- (a) Similarly as previous one:

In [4]:
b1 = 20
b2 = 21
b3 = 22
r1 = 0.06
r2 = 0.065
r3 = 0.07
PV = b1*(1+r1)**-1 + b2*(1+r2)**-2 + b3*(1+r3)**-3

print(f'The present value of the cash flows is ${PV:.2f}')
x = PV*((1+r1)**-1 + (1+r2)**-2 + (1+r3)**-3)**-1
print(f'The three year swap price is ${x:.2f}')

The present value of the cash flows is $55.34
The three year swap price is $20.95


- (b): Thus, we repeat the process without the first year settlement.

In [5]:
PV_new = b2*(1+r2)**-2 + b3*(1+r3)**-3
x_new = PV_new*((1+r2)**-2 + (1+r3)**-3)**-1
print(f'The new swap price is ${x_new:.2f}')


The new swap price is $21.48


## Exercise 8.3
Consider the same 3-year oil swap. Suppose a dealer is paying the fixed price and
receiving floating. What position in oil forward contracts will hedge oil price risk in
this position? Verify that the present value of the locked-in net cash flows is zero.

### Solution:
Since the dealer is paying fixed and receiving floating, she generates the cash-flows depicted in
column 2. Suppose that the dealer enters into three short forward positions, one contract for each
year of the active swap. Her payoffs are depicted in column 3, and the aggregate net cash flow
position is in column 4.

In [6]:
data = {
    "Year": [1, 2, 3],
    "Net Swap Payment": ["S₁ - $20.9519", "S₁ - $20.9519", "S₁ - $20.9519"],
    "Short Forwards": ["$20 - S₁", "$21 - S₁", "$22 - S₁"],
    "Net Position": [-0.9519, 0.0481, 1.0481]
}

df = pd.DataFrame(data)
df

Year Net Swap Payment Short Forwards  Net Position
0     1    S₁ - $20.9519       $20 - S₁       -0.9519
1     2    S₁ - $20.9519       $21 - S₁        0.0481
2     3    S₁ - $20.9519       $22 - S₁        1.0481

We need to discount the net positions to year zero

In [7]:
p1 = -0.9519
p2 = 0.0481
p3 = 1.0481
PV = p1*(1+r1)**-1 + p2*(1+r2)**-2 + p3*(1+r3)**-3
print(f'The present value of the net position is ${PV:.2f}')

The present value of the net position is $-0.00


## Exercise 8.4
Consider the 3-year swap in the previous example. Suppose you are the fixed-rate
payer in the swap. How much have you overpaid relative to the forward price after
the first swap settlement? What is the cumulative overpayment after the second swap
settlement? Verify that the cumulative overpayment is zero after the third payment.
(Be sure to account for interest.)

### Solution:
The fair swap rate was determined to be $20.952. Therefore, compared to the forward curve
price of $20 in one year, we are overpaying $0.952. In Year 2, this overpayment has increased to
$0.952 × 1.070024 = 1.01866, where we used the appropriate forward rate to calculate the
interest payment. In Year 2, we underpay by $0.048, so that our total cumulative underpayment
is $0.97066. In Year 3, this overpayment has increased again to $0.97066 × 1.08007 = 1.048.
However, in Year 3, we receive a fixed payment of 20.952, which underpays relative to the
forward curve price of $22 by $22 − $20.952 = 1.048. Therefore, our cumulative balance is
indeed zero, which was to be shown.

## Exercise 8.5
Consider the same 3-year swap. Suppose you are a dealer who is paying the fixed
oil price and receiving the floating price. Suppose that you enter into the swap and
immediately thereafter all interest rates rise 50 basis points (oil forward prices are
unchanged). What happens to the value of your swap position? What if interest rates
fall 50 basis points? What hedging instrument would have protected you against
interest rate risk in this position?

### Solution:
Now we compute the new swap price based on the change of the interest rate:

In [8]:
delta = 0.005
PV_raise = b1*(1+r1+delta)**-1 + b2*(1+r2+delta)**-2 + b3*(1+r3+delta)**-3
PV_fall = b1*(1+r1-delta)**-1 + b2*(1+r2-delta)**-2 + b3*(1+r3-delta)**-3
print(f'The present value of the cash flows if rates rise is ${PV_raise:.2f}')
print(f'The present value of the cash flows if rates fall is ${PV_fall:.2f}')

x_raise = PV_raise*((1+r1+delta)**-1 + (1+r2+delta)**-2 + (1+r3+delta)**-3)**-1
x_fall = PV_fall*((1+r1-delta)**-1 + (1+r2-delta)**-2 + (1+r3-delta)**-3)**-1
print(f'The new swap price if rates rise is ${x_raise:.3f}')
print(f'The new swap price if rates fall is ${x_fall:.3f}')

PV_net_raise = p1*(1+r1+delta)**-1 + p2*(1+r2+delta)**-2 + p3*(1+r3+delta)**-3
PV_net_fall = p1*(1+r1-delta)**-1 + p2*(1+r2-delta)**-2 + p3*(1+r3-delta)**-3
print(f'The present value of the net position if rates rise is ${PV_net_raise:.2f}')
print(f'The present value of the net position if rates fall is ${PV_net_fall:.2f}')

The present value of the cash flows if rates rise is $54.83
The present value of the cash flows if rates fall is $55.86
The new swap price if rates rise is $20.949
The new swap price if rates fall is $20.955
The present value of the net position if rates rise is $-0.01
The present value of the net position if rates fall is $0.01


The dealer could have tried to hedge his exposure with a forward rate agreement or any other
derivative protecting against interest rate risk.

## Exercise 8.6
Supposing the effective quarterly interest rate is 1.5%, what are the per-barrel swap
prices for 4-quarter and 8-quarter oil swaps? (Use oil forward prices in Table 8.9.)
What is the total cost of prepaid 4- and 8-quarter swaps?

### Solution:
We use the information on the table and redo the procedure again.

In [ ]:
cashflow = [21, 21.1, 20.8, 20.5, 20.2, 20, 19.9, 19.8]
#coupons = [0.9852, 0.9701, 0.9546, 0.9388, 0.9231, 0.9075, 0.8919, 0.8763]
#r = [1/c - 1 for c in coupons]
r = [0.015]*len(cashflow)
PV4 = sum(cf*(1+r[i])**-(i+1) for i, cf in enumerate(cashflow[:4]))
PV8 = sum(cf*(1+r[i])**-(i+1) for i, cf in enumerate(cashflow))
x4 = PV4*((1+r[0])**-1 + (1+r[1])**-2 + (1+r[2])**-3 + (1+r[3])**-4)**-1
x8 = PV8*((1+r[0])**-1 + (1+r[1])**-2 + (1+r[2])**-3 + (1+r[3])**-4 + (1+r[4])**-5 + (1+r[5])**-6 + (1+r[6])**-7 + (1+r[7])**-8)**-1
print(f'The total present value of the cash flows is ${PV4:.2f}')
print(f'The total present value of the cash flows is ${PV8:.2f}')
print(f'The four quarter swap price is ${x4:.2f}')
print(f'The eight quarter swap price is ${x8:.2f}')

The total present value of the cash flows is $80.38
The total present value of the cash flows is $152.93
The four quarter swap price is $20.85
The eight quarter swap price is $20.43


## Exercise 8.7
Using the information about zero-coupon bond prices and oil forward prices in Table
8.9, construct the set of swap prices for oil for 1 through 8 quarters.

### Solution:
Let the forward on the settlement day to be $F_{0,t}$. Then amount we will get paid is:
$$\sum_{i=0}^n F_{0,t_i}P(0,t_i)$$
Then the swap price satisifies:
$$\sum_{i=0}^n R P(0,t_i) = \sum_{i=0}^n F_{0,t_i}P(0,t_i)$$
Thus, we can solve:
$$ R=\frac{\sum_{i=1}^n P\left(0, t_i\right) F_{0, t_i}}{\sum_{i=1}^n P\left(0, t_i\right)}$$

In [14]:
coupons = [0.9852, 0.9701, 0.9546, 0.9388, 0.9231, 0.9075, 0.8919, 0.8763]
paid = [cash*coupons[i] for i, cash in enumerate(cashflow)]
R = []
for i in range(len(paid)):
    R.append(sum(paid[:i+1])/sum(coupons[:i+1]))
print(f'The swaps are ${R}')
    

The swaps are $[21.0, 21.049613869994378, 20.967727413313174, 20.853636292774183, 20.72719099710801, 20.610992551899002, 20.51449050401753, 20.430420946626384]


## Exercise 8.8
Using the information in Table 8.9, what is the swap price of a 4-quarter oil swap
with the first settlement occurring in the third quarter?

### Solution:
Now in this case we only have 3 periods. The formula is
$$ X=\frac{\sum_{i=3}^6 P_0\left(0, t_i\right) F_{0, t_i}}{\sum_{i=3}^6 P_0\left(0, t_i\right)}$$

In [11]:
print(f'The swaps are ${sum(paid[2:6])/sum(coupons[2:6]):.2f}')

The swaps are $20.38


## Exercise 8.9
Given an 8-quarter oil swap price of $20.43, construct the implicit loan balance for
each quarter over the life of the swap.

### Solution:
Using the eight-quarter swap price of $20.43, we can calculate the net position by subtracting the
swap price from the forward prices. The one-quarter implied forward rate is calculated from the
zero-coupon bond prices. The column implicit loan balance adds the net position of each quarter
and the implicit loan balance plus interest of the previous quarter.

In [12]:
swap = R[-1]
net = [c - swap for c in cashflow]
print(f'The net position is ${net}')
# Implied forward rate
r = [1/coupons[0]]
for i in range(1, len(coupons)):
    r.append((coupons[i-1]/coupons[i]))
print(f'The implied forward rate is {r}')
# Balance
balance = [net[0]]
for i in range(1, len(net)):
    balance.append(balance[i-1]*(r[i]) + net[i])
print(f'The balance is ${balance}')

The net position is $[0.5695790533736158, 0.6695790533736172, 0.3695790533736165, 0.0695790533736158, -0.2304209466263849, -0.4304209466263842, -0.5304209466263856, -0.6304209466263835]
The implied forward rate is [1.0150223304912709, 1.0155654056282857, 1.016237167399958, 1.0168299957392417, 1.0170079081356298, 1.017190082644628, 1.0174907500840902, 1.017802122560767]
The balance is $[0.5695790533736158, 1.2480238357503683, 1.6378672610642013, 1.7350116134629712, 1.5340995849726153, 1.130049936997, 0.6193944114011709, 5.218048215738236e-15]


We have 0 in the end.